### ISSUE TITLE
Excel class API changes

### ISSUE THREAD

**Issue by collinsmith** [collinsmith/riiablo#26]
I want to look into changing some functionality of `Excel` and improving it. Some of these are tentative

- [x] remove reliance on TXT/optimize -- TXT isn't used directly anywhere, so I want to look into if any optimizations can be made in the excel loading if I don't have to go through TXT first, however it might be the case that `TXT` can be interchangeable with `BIN` when that is added
- [x] add support for `BIN` files which can be parsed faster than `TXT` and contain a subset of columns (namely removing the metadata columns not needed for a release version)
- [ ] foreign key support -- flag columns as foreign keys which can be assigned at a later time (second pass of the parser once the other tables have loaded). This is an extremely common use-case, so it might improve readability if this were the case that I could simply replace these columns with references to the corresponding data, instead of the index of that data. I want to avoid tightly coupling the tables with `Riiablo.files` if I can though.
- [ ] transition from cascading if-statements to adding a map indexed on type with parsers for the excel columns. I think this will improve speed and extensibility, but I also want to avoid auto-boxing the data wherever possible
- [ ] it might be nice to support parsing data structures, e.g., some excel files contain multiple numbered columns (`stat1, param1, stat2, param2`), but I think it might be nice if this could be parseable into `stat[]` object consisting of a `stat, param` instead of `stat[], param[]`, this just keeps associated data together
- [ ] related to above point, it would be nice to support multi-dimensional arrays, or at least 2D. E.g., some stats have copies for each difficulty, and partial set bonuses would be easier to implement in a 2D array.

This might be doable when some #8 improvements on loading TXT files are made.

**Comment by collinsmith** [comment]
Another good use-case for 2D arrays -- `quests.txt` has 3 indexes for each `qsts`.

**Comment by collinsmith** [comment]
Relating to #8, I wrote a quick benchmark testing the read speeds of `monstats.txt` on Android using different methods, and it seems much of the problem lies within current `TXT` parsing. `TXT` indicates loading and parsing the file, while `bytes` indicates reading the file handle into a byte array.

||internal|mpq|
|---|---|---|
|txt|214 ms|245 ms|
|bytes|5 ms|45 ms|

**Comment by collinsmith** [comment]
I started writing my own (de)serializers for `monstats`, and preliminary tests look extremely promising. Down to `225ms` from `4669ms` or about 95% reduction. It seems pretty obvious this is the way to go, and there are still some fairly significant optimizations that can be made in terms of excluding unneeded columns and storing certain strings as index references (e.g., string table references can be saved as the actual string table index rather than reference string).

The plan I'm working with right now is to generate my own bin files for (at least some) txt excel files. I'm still working out the generation process and figuring out if this should be supported by all files*. I wrote a code generation tool to perform most of the heavy lifting on `monstats` and any other excel file (spits out the read/write code to be copy/pasted into the source code), however I still need to test equality of txt excel objects versus bin excel objects to make sure that the serialization process is performed correctly (basic tests look fine though).

*The majority of the slowdown is due to a few specific files, so I want to see the effect of fixing those first, but eventually I'll target all files.

**Comment by collinsmith** [comment]
I'd like to push some of the (de)serialization responsibilities up to `Excel` from the subclasses, as obviously an `int` representing the number of rows of the table should always come first. Additionally, `MonStats` deserialization is taking care of the indexing requirements that should also probably be handled by `Excel`. The deserialization process should index exactly how the txt tables do, but I'm concerned that if `Excel` handles it, it could be really slow if I need to access the txt table at all, which doesn't necessarily indicate the order of which the columns are serialized. One possible solution would be to guarantee the first sequence of bytes are always the index (`String` or `int`)

**Comment by collinsmith** [comment]
I rewrote the excel system under `com.riiablo.excel`, but this ended up being more work than I'd like to do at once. I'm going look into retrofitting some of the changes I made onto `com.riiablo.codec.excel`, but I think I'm going to go with the reflection and annotation approach I initially mentioned above and when I want to come back and improve this system later, implement the much better approach I began in `com.riiablo.excel`. Additionally, support for serializing some specific excels needs to be looked into, some classes like `Runes` and `BodyLocs` do some re-indexing that should be specifically tested.

<details>
<summary>com.riiablo.excel spec</summary>

```java
public abstract class Excel<T extends Excel.Entry, U extends Serializer<T>> implements Iterable<T> {
  public static <E extends Entry, S extends Serializer<E>, T extends Excel<E, S>>
  T load(T excel, FileHandle txt) {
    return load(excel, txt, null);
  }

  public static <E extends Entry, S extends Serializer<E>, T extends Excel<E, S>>
  T load(T excel, FileHandle txt, FileHandle bin) {...}
}

public interface Serializer<T extends Excel.Entry> {
  void readBin(T entry, DataInput in) throws IOException;
  void writeBin(T entry, DataOutput out) throws IOException;
}

public class MonStats extends Excel<MonStats.Entry, MonStats.Serializer> {
  public MonStats() {
    super(Entry.class);
  }

  @Override
  public Entry newEntry() {
    return new Entry();
  }

  @Override
  public Serializer newSerializer() {
    return new Serializer();
  }

  public static class Entry extends Excel.Entry {}

  public static class Serializer implements com.riiablo.excel.Serializer<Entry> {
    @Override public void readBin(Entry entry, DataInput in) {}
    @Override public void writeBin(Entry entry, DataOutput out) {}
  }
}

FileHandle txt = ...;
FileHandle bin = ...;
MonStats excel = Excel.load(new MonStats(), txt, bin);
```
</details>

**Comment by collinsmith** [comment]
I rewrote `TXT` as `TxtParser` to be a wrapper for a `BufferedReader`. This gives slightly better performance than `TXT` (on mobile) and removes much of the heavy-weightedness. Excels are now parsed line-by-line with `EXPANSION` row being automatically ignored -- this change results in only the actively read row's data being accessible by `Excel.loadTxt` (which is fine -- `TXT` was written to fit this use-case, but it isn't needed anymore).

**Comment by collinsmith** [comment]
I've been rewriting this module using techniques I've learned since creating it, since tons of the info in this is outdated and deprecated. The big improvement I've made is while rewriting `TxtParser`, I am using more `byte`-oriented data structures and readers and enforcing ASCII characters. No more expensive `readLine()` or `String#split` operations -- instead the `InputStream` is read until certain bytes are found (`\t`, `\r`, `\n`, `-1`), caching their locations and then the data can be parsed in-line using this buffer directly. Strings are returned as `AsciiString` references to the buffer and `String` is only generated when needed.

Exception handling is significantly more robust and error reporting far improved. I'm moving forward with final tests and working on the `BinGenerator` so I can merge and close this issue. The next iteration to use annotation processing will likely happen at a later time.
